runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,601 @@
+        context.getDataSetOrThrow(stagePayload.getInput());
+
+    MapPartitionOperator<WindowedValue<InputT>, RawUnionValue> taggedDataset =
+        new MapPartitionOperator<>(
+            inputDataSet, typeInformation, function, transform.getUniqueName());
+
+    for (SideInputId sideInputId : stagePayload.getSideInputsList()) {
+      String collectionId =
+          components
+              .getTransformsOrThrow(sideInputId.getTransformId())
+              .getInputsOrThrow(sideInputId.getLocalName());
+      // Register under the global PCollection name. Only ExecutableStageFunction needs to know the
+      // mapping from local name to global name and how to translate the broadcast data to a state
+      // API view.
+      taggedDataset.withBroadcastSet(context.getDataSetOrThrow(collectionId), collectionId);
+    }
+
+    for (String collectionId : outputs.values()) {
+      pruneOutput(
+          taggedDataset,
+          context,
+          outputMap.get(collectionId),
+          (Coder) outputCoders.get(collectionId),
+          transform.getUniqueName(),
+          collectionId);
+    }
+    if (outputs.isEmpty()) {
+      // NOTE: After pipeline translation, we traverse the set of unconsumed PCollections and add a
+      // no-op sink to each to make sure they are materialized by Flink. However, some SDK-executed
+      // stages have no runner-visible output after fusion. We handle this case by adding a sink
+      // here.
+      taggedDataset.output(new DiscardingOutputFormat());
+    }
+  }
+
+  private static <T> void translateFlatten(
+      String id, RunnerApi.Pipeline pipeline, BatchTranslationContext context) {
+    Map<String, String> allInputs =
+        pipeline.getComponents().getTransformsOrThrow(id).getInputsMap();
+    DataSet<WindowedValue<T>> result = null;
+
+    if (allInputs.isEmpty()) {
+
+      // Create an empty dummy source to satisfy downstream operations. We cannot create an empty
+      // source in Flink, so we send the DataSet to a flatMap that never forwards its element.
+      DataSource<String> dummySource = context.getExecutionEnvironment().fromElements("dummy");
+      result =
+          dummySource
+              .<WindowedValue<T>>flatMap(
+                  (s, collector) -> {
+                    // never return anything
+                  })
+              .returns(
+                  new CoderTypeInformation<>(
+                      WindowedValue.getFullCoder(
+                          (Coder<T>) VoidCoder.of(), GlobalWindow.Coder.INSTANCE)));
+    } else {
+      for (String pCollectionId : allInputs.values()) {
+        DataSet<WindowedValue<T>> current = context.getDataSetOrThrow(pCollectionId);
+        if (result == null) {
+          result = current;
+        } else {
+          result = result.union(current);
+        }
+      }
+    }
+
+    // Insert a dummy filter. Flink produces duplicate elements after the union in some cases if we
```

Comment:
<!--new_thread; commit:c3bc9f1ff4d5f830b19aef83b758b12e363f7679; resolved:0-->
If true, this seems like a really, really bad bug in Flink. Is it tracked somewhere?